### PR TITLE
iOS support

### PR DESCRIPTION
**Pull request by AJenbo**
This PR adds initial support for iPhone and other iOS devices. This is still a basic port as nothing has been adapted to the mobile form factor or events.

At the moment network play has been disabled, to get this running lib sodium has to be added to the build process.

Todo:
- [ ] Make the project use relative paths
- [ ] Build for ARM CPUs
- [ ] Merge @richstoner's changes https://github.com/diasurgical/devilutionX/pull/182

### CODE REVIEW

**Comment by richstoner** [comment]
This is great. Took some tweaks to get it playable in a simulator, but can confirm working. Took another 2 hours to get this building & running on a device.

Good news is that the on-device runs. Bad news is that it runs into the same color / 64-bit playability issues that the OSX 64-bit builds have. I suspect if we fix that, we'll have a running version on iOS 12+.

**Comment by AJenbo** [comment]
@richstoner could you post the changes you did. I don't own a iPhone or Mac so I only get to work on it when I can borrow a Mac and can only test it out on the simulator. But thought it would be good to post it early in case others where able to contribute since I got it working on a simulator. I also know little about developing for Apple ecosystems or using Xcode :D

Then I can switch focus to fixing color issue. My frind how helped me getting in to Xcode did report it the other day but we whereat sure if it was something with his machine or a general issue.

**Comment by AJenbo** [comment]
@richstoner I just tested on a machine running macOS 10.13.6 and colors appeared correctly. Did you use the cmake build process (as per the readme instructions) or the xcode project?

**Comment by richstoner** [comment]
@AJenbo Will do. Currently using the xcode project in Xcode 10.2.1. I'll try the cmake build and report back. 

This is how the colors appear on device (and the Mac 64bit build from xcode)

Screen Shot 2019-07-06 at 8 37 47 AM

The main menu and loading screen colors look fine.

I'll also start on cleaning up the changes I made and do a proper fork.

**Comment by richstoner** [comment]
@AJenbo - You were correct. The 64bit cmake build on Mac has correct color. I'm probably missing a compilation flag somewhere in the stack. Shouldn't be too hard to track down.

**Comment by AJenbo** [comment]
Maybe @BennyFranco can be helpful in regards to the Xcode project producing wrong colors.

**Comment by BennyFranco** [comment]
> @AJenbo Will do. Currently using the xcode project in Xcode 10.2.1. I'll try the cmake build and report back.
> 
> This is how the colors appear on device (and the Mac 64bit build from xcode)
> 
> Screen Shot 2019-07-06 at 8 37 47 AM
> 
> The main menu and loading screen colors look fine.
> 
> I'll also start on cleaning up the changes I made and do a proper fork.

I had this issue before, only running the game attached to llvm using Xcode 10, but if run the game without attach it to Xcode runs normally, we'll check it.

**Comment by richstoner** [comment]
Some additional info:

64bit cmake build with brew bundle -> working on Mac
64bit cmake build with SDL, SDL_ttf, SDL-mixer build and installed locally -> working on Mac
64bit xcode build with SDL, SDL_ttf, SDL-mixer built with xcode -> color issue
64bit xcode build with SDL, SDL_ttf, SDL-mixer built with xcode -> color issue

Working on a root cause.

**Comment by BennyFranco** [comment]
> Some additional info:
> 
> 64bit cmake build with brew bundle -> working on Mac
> 64bit cmake build with SDL, SDL_ttf, SDL-mixer build and installed locally -> working on Mac
> 64bit xcode build with SDL, SDL_ttf, SDL-mixer built with xcode -> color issue on Mac
> 64bit xcode build with SDL, SDL_ttf, SDL-mixer built with xcode -> color issue on iOS
> 
> Working on a root cause.

Captura de Pantalla 2019-07-07 a la(s) 18 18 31

Hi @richstoner! be sure that the option in the picture is disabled, Xcode adds the argument
`--NSDocumentRevisionsDebugMode YES`, we don't have a document based application so, it's safe disable it. Maybe you can update the Xcode schemes and include it with your changes. What do you think about it @AJenbo ?

**Comment by AJenbo** [comment]
I'm not entirely sure what this option does but yeah let's disable it if it causes issues and doesn't give us any benefits.

**Comment by richstoner** [comment]
@BennyFranco - that resolved the issue on the Mac xcode build. I was also able to fix the Mac issue by building for release. 

The iOS project doesn't have the NSDocumentRevisionsDebugMode flag, but I suspect it's related to a debugging flag.

**Comment by AJenbo** [comment]
The only thing I can think of that could cause this is the following setting, but I don't see how that might have been enabled:
https://github.com/diasurgical/devilutionX/blob/73a29f0e68b2ff9f23a319f747f45a905376ddf0/defs.h#L97

**Comment by richstoner** [comment]
Actually, forget what I said above - this might be the diablo command line parser issue. I had to override some of the default settings to pass the sandboxed Documents path.

**Comment by richstoner** [comment]
Success! Playable on device. This code is the culprit: https://github.com/diasurgical/devilutionX/blob/master/Source/diablo.cpp#L321-L466

It must be interpreting the command line args incorrectly (including the '''--NSDocumentRevisionsDebugMode YES''' argument) and hitting one of the debug modes. Simply commenting this out fixed the color issues. 

What's the best way for me to roll this back in? I'll try to get this into a fork sometime this week. 

The current setup:
1) Build, sign, and install the app on device
2) Copy the diabdat.mpq via itunes file sharing
3) Run!

**Comment by richstoner** [comment]
@AJenbo @BennyFranco 

code & readme here : https://github.com/richstoner/devilutionX/tree/iOS-arm64

Thanks for the help!

**Comment by mdstage** [comment]
Hi everyone
@richstoner about the step "Copy the diabdat.mpq via itunes file sharing"
in the iOS build, the path to diabdat.mpq is `./diabdat.mpq`
What about in method
```
// init.cpp
HANDLE init_test_access(char *mpq_path, char *mpq_name, char *reg_loc, int flags, int fs)
```
to add a line with `SFileOpenArchive` testing the path at the root of the project? (`./diabdat.mpq`)
Downside is that it's modifying main code of DevilutionX instead of just the iOS "wrapper" with configs.

With this change I was able to run the build of this branch (did not test yet the fork of @richstoner) on iPad simulator (all fine), on iPhone simulator (with glitches on the lower part of the screen), on iPad device (all fine), but not on iPhone device (some SDL crashes).

**Comment by richstoner** [comment]
Hey @mdstage - on your iPad device build, is the mpq file writable? That was my reason for moving it out of the bundle. 

Ignore my fork for now. Most of the changes are superfluous and can probably be condensed into a single modification as you propose.

**Comment by mdstage** [comment]
@richstoner I m not sure what you mean by writable, is it edited in some way by the code, no read only?
What I can say is that the app is running on iPad device.
To test, just for one build, you can add at the end of init_test_access:
```
HANDLE init_test_access(char *mpq_path, char *mpq_name, char *reg_loc, int flags, int fs)
{
...
        if (SFileOpenArchive("diabdat.mq", flags, fs, &archive))
                return archive;
}
```
This piece of code is really only to do a quick test on this file loading (its ignoring the parameters of init_test_access etc.)
If this approach is used, it should be done in a clean way.

**Comment by AJenbo** [comment]
See https://github.com/diasurgical/devilutionX/issues/45 for the read/write issue.

**Comment by mdstage** [comment]
I see. Thanks for the answer. It seems that it is writable, as devilutionX is starting correctly.

**Comment by tomkidd** [comment]
What's the status on this, or mobile/touch in general?

With some work I've got it up and running in Xcode for iOS devices, though things like the touchscreen aren't doing anything. Reading through this and some other threads it seems likely that the reason I got it going so quickly is because folks here have worked through a lot of the issues with graphics, sound, arm64, etc. already. 

I've been able to get a few old games running on iOS/tvOS, if I can be of help let me know.

**Comment by AJenbo** [comment]
@tomkidd are you referring to getting the master branch running?

Yes, we have cleaned things up and made the code significantly more portable since the initial work on getting it running on iOS.

Touch is now handled explicitly via `SourceX/controls/touch.cpp`, also mouse emulation was disabled by default recently when merging the PlayStation Vita port as it doesn't make sense there: `SourceX/display.cpp:123`

Personally, my problem with producing an iOS port is that I do not have any relevant Apple hardware which appears to be a requirement.
For this PR I borrowed a MacBook Pro for a weekend which was not enough time to figure out how to get builds working based on CMake, having builds based on CMake is practically a requirement as else it will break the build whenever we make general changes. 

Another issue is how to build and distribute releases. What I got working required that you had the game assets available at build time and thus prevented distribution, making for a very non-ideal situation.

Lastly, mouse emulation is rather non-ideal a much better user experience would be to implement a virtual gamepad and leverage the general controller code that we have for controls.

In short, if you can provide the needed changes for getting it building for iOS via CMake (generating an Xcode project and having to do a couple of steps would be acceptable) then that would be good enough to merge and finally close this PR.

**Comment by tomkidd** [comment]
Interesting, yeah I can take a stab at it. I'm not that well versed with CMake anymore but I could probably get it going. 

And yeah the work I got going was from the master branch. So some of the things people noted like sounds and color issues I wasn't getting since I had the latest hotness.

Touch screen support might be a good thing to have for menus and inventory and so forth, maybe less so for the actual gameplay. In some of the id Tech ports I've done I've put a virtual joystick on the screen with UIKit but it's not ideal because it's got some threading issues so we'd need to explore alternatives.

**Comment by Blacklion55** [comment]
> > @AJenbo Will do. Currently using the xcode project in Xcode 10.2.1. I'll try the cmake build and report back.
> 
> > 
> 
> > This is how the colors appear on device (and the Mac 64bit build from xcode)
> 
> > 
> 
> > Screen Shot 2019-07-06 at 8 37 47 AM
> 
> > 
> 
> > The main menu and loading screen colors look fine.
> 
> > 
> 
> > I'll also start on cleaning up the changes I made and do a proper fork.
> 
> 
> 
> I had this issue before, only running the game attached to llvm using Xcode 10, but if run the game without attach it to Xcode runs normally, we'll check it. 

Hahsnb22289

**Comment by AJenbo** [comment]
Things are now mostly running but I need help with testing on a real device and some minor issues, please see this PR: #3612
